### ISSUE TITLE
Support narrowing on tuple match cases

### ISSUE THREAD

**Issue by MatthewMckee4** [astral-sh/ty#561]
From the typing spec: https://typing.python.org/en/latest/spec/tuples.html#type-compatibility-rules, we currently do not support this.
```py
def func(subj: tuple[int | str, int | str]):
    match subj:
        case x, str():
            reveal_type(subj)  # tuple[int | str, str]
        case y:
            reveal_type(subj)  # tuple[int | str, int]
```

**Comment by carljm** [comment]
Another slightly different case, from #2708:

```py
from typing import assert_never

def func(value: int | tuple[int, int]) -> int:
    match value:
        case int(value):
            return value
        case (left, right):
            return left + right
        case _:
            assert_never(value)  # this assert should pass, currently we still see `tuple[int, int]` here
```

In this case we are not narrowing the types of particular elements, but since we are capturing all two-element tuples, we should narrow the match subject type accordingly.

**Comment by carljm** [comment]
@charliermarsh this is assigned to you -- are you still working on it?

**Comment by charliermarsh** [comment]
I think this needs https://github.com/astral-sh/ty/issues/3314 in order to support cases involving `Sequence`, otherwise we fail to prove that `Sequence[int] & str` are disjoint in various cases.
